pymeasure/instruments/hcp/tc038.py

```diff
@@ -0,0 +1,135 @@
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+
+from pymeasure.instruments import Instrument
+from pyvisa.constants import Parity
+
+
+def _data_to_temp(data):
+    """Convert the returned hex value "data" to a temperature in °C."""
+    return int(data[7:11], 16) / 10.
+    # get the hex number, convert to int and shift the decimal sign
+
+
+registers = {'temperature': "D0002",
+             'setpoint': "D0120",
+             }
+
+
+class TC038(Instrument):
+    """
+    Communication with the HCP TC038 oven.
+
+    This is the older version with an AC power supply and AC heater.
+
+    It has parity or framing errors from time to time. Handle them in your
+    application.
+    """
+
+    def __init__(self, resourceName, address=1, timeout=1000,
+                 includeSCPI=False):
+        """
+        Initialize the communication.
+
+        Parameters
+        ----------
+        resourceName : str
+            name COM-Port.
+        address : int
+            address of the device. Should be between 1 and 99.
+        timeout : int
+            Timeout in ms.
+        """
+        super().__init__(resourceName, "TC038", timeout=timeout,
+                         write_termination="\r", read_termination="\r",
+                         parity=Parity.even,)
+        self.address = address
+
+        self.set_monitored_quantity()  # start to monitor the temperature
+
+    def _adjust_command(self, command):
+        """
+        Return an adjusted command string for the oven protocol.
+
+        :param str command: Three chars indicating the type, and data for
+            the command, if necessary.
+        """
+        # 010 is CPU (01) and time to wait (0), which are fix
+        return chr(2) + f"{self.address:02}" + "010" + command + chr(3)
+        # Response is chr(2) + address:02 + "01" + response + chr(3)
+
+    def write(self, command):
+        """Send a "command" in its own protocol and clear the buffer."""
+        # The oven responds, use therefore ask to clear the buffer.
+        self.ask(command)
```

Comment:
watch out, an  `ask` is (effectively) a write+read; this should be afaict `super().write(self._adjust_command(command)))`

Reply:
I know. The problem is, that the device responds to a change. Therefore, if I set a variable, an acknowledgment will sit in the buffer. For the next query, this acknowledgement is returned instead of the desired response.

The setters invoke write, and therefore the buffer has to be cleared afterwards. Maybe I could write "clear the read buffer" and hint to another (write_only or so) method, which does not read.

Reply:
If I don't misunderstand, the `check_set_errors`/`check_get_errors` toggles are there for exactly that behaviour -- you implement an overridden [`check_errors`](https://pymeasure.readthedocs.io/en/latest/api/instruments/instruments.html#pymeasure.instruments.Instrument.check_errors) method (base one [here](https://github.com/pymeasure/pymeasure/blob/master/pymeasure/instruments/instrument.py#L514)) that checks such an acknowledgement (and also responds when it is not "OK").

We should probably extend [the docs](https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments.html#writing-properties) to explain more about the various options?

Reply:
Oh, I did not think about using error checking as a read method. Documentation is geared towards SCPI and I did not want to "request" the errors. Nevertheless, the documentation ist good enough, that I directly understood how to make it work. It's just a task to get into such a framework and all of its possibilities.

It is not optimal, because the same check_errors can not be applied to reading, but nevertheless, it works.
Now a meaningful error message is given for the ConnectionError.